### ISSUE TITLE
MIDI output timing is off

### ISSUE THREAD

**Issue by mankeli** [schismtracker/schismtracker#75]
It seems that timing of the midi output (to alsa) is jittering badly. I'm using Linux and using different audio drivers don't seem to have any effect on the issue.

You can easily test this by running Timidity (or other midi-inputting synth) and outputting a straight sequence of notes to it. The sounds should come out precisely, but instead their timing is quite random sometimes and it seems to be affected by the cpu load.

**Comment by jangler** [comment]
Seems like MIDI was not well implemented, since we're seeing so many bugs with it… I wish I had better domain knowledge for this. Maybe @mb14 might know something about it?

**Comment by ghost** [comment]
Yes I get the feeling MIDI was a bit of an afterthought (...and rightly so, I suppose). Sadly I don't have any MIDI output devices or use Linux. I'm not sure what to recommend here but maybe if you can find other users who are having similar problems it'd help pin it down a bit? Sorry. Thanks for the bug report!

**Comment by mankeli** [comment]
@mb14 can you try with your system's internal midi synth?

**Comment by esaruoho** [comment]
@mankeli I'm giving it a shot with just a regular pattern of 16 rows with a GM internal midi synth bassdrum, triggered four times, but with a C00 on the 15th row of the pattern - and there's a yank every time we get to the end of the pattern - i.e., something off with the timing. I'm running on a windows10 and newest schismtracker build. at least with speed 04/128, there's a yank. but if i switch to 06/128, there's no yank.

however, if i'm at 06/128 and have 16 rows of midinotes triggering a piano, there's a certain shuffle and it really does not retain any kind of tempo whatsoever - it's like schism isn't sending midi out very stably at all. it's like there's a random note delay offset on there at all times.

![glinksch](https://cloud.githubusercontent.com/assets/4966687/25267839/19d4790a-267f-11e7-9702-516ea28abf62.png)

**Comment by mankeli** [comment]
@esaruoho This sounds like the behavior I get on Linux too. I suspect that the midi output has been added in a wrong place, and thus not outputting notes with exact timing, but I haven't found the code for this yet.

**Comment by vcaputo** [comment]
@mankeli if you could provide me with specific instructions for getting timidity + schismtracker reproducing what you're experiencing, I'll take a look at fixing it.   I'm on debian jessie x86_64, there's a timidity and timidity-daemon package but I don't know what's involved in getting schismtracker to make noise through timidity, I'm a coder not (yet) a musician.

**Comment by vcaputo** [comment]
Nevermind, I've got timidity making noise through schism, reproduced the problem.

**Comment by vcaputo** [comment]
I'm not seeing any of these problems playing some test patterns through timidity, but I do experience very frustratingly unreliable interactive note playing which I've hacked together a patch to fix.  I wonder how much of the complaints are surrounding that issue, as it makes things seem to be totally random and unpredictable, even laggy-appearing.

**Comment by jangler** [comment]
Yeah, playing a regular pattern through timidity is totally stable for me using the latest master regardless of speed and tempo, although there is a lot of latency.

I look forward to testing #93, since it's definitely annoying to track with MIDI as-is.

**Comment by jangler** [comment]
Leaving this open for now since I get about a half a second of latency using Schism through TiMidity. Is it this bad for everyone else, too?

**Comment by vcaputo** [comment]
Hmm, I don't see **that much** latency, seems more like ~100ms - but it's pretty much identical to the latency I observe running `vkeybd --addr 128:0`.  Have you tried `vkeybd` to see if it's specific to schism?

**Comment by vcaputo** [comment]
After experimenting a bit more with some fairly complex songs where I've reconfigured their instruments to trigger MIDI channels instead, letting the songs play to completion and loop a few times seems to slowly leak midi channels - even pressing F8 doesn't stop their sustained sounds, and eventually between timidity and schism my CPUs are bogged down and everything gets pretty crazy sounding due to the audio buffer underruns.  But this is only after a number of minutes of looping these songs.  I'm not sure if that's related to this issue or warrants opening a new one.

poor old timidity...
```
  PID USER      PR  NI    VIRT    RES    SHR S %CPU %MEM     TIME+ COMMAND                                                                                              
10479 timidity  20   0   99548  12964   6012 R 99.9  0.3  32:15.87 /usr/bin/timidity -Os -iAD          
``` 

killing schism is all that's needed to restore timidity's sanity, definitely a schism midi bug.

Addendum:
timidity is proving to be a quite useful tool in debugging schism's MIDI integration.
It appears I can prevent the degenerate behavior by running timidity with the `--no-overlap-voice` flag, though the playback quality is of course harmed.  What seems to be happening is schism leaks some notes and they're accumulating with every repeat of the song, eventually overloading timidity's mixing engine.

**Comment by vcaputo** [comment]
@jangler WRT latency w/timidity, I found it to be relatively low with the defaults but things became practically latency-free after adding `-q0/0 -B2,8` to my timidity flags.  You can also run timidity with --realtime-priority=N to make things even better - but that's probably not the wisest move until schism stops leaking overlapping notes because timidity eventually hogs 100% of a cpu.

**Comment by jangler** [comment]
I would open a new issue for the MIDI note leaks. I'll take a look at the latency stuff.

**Comment by jangler** [comment]
OK, the latency is also present using vkeybd and disappears when I use those flags with TiMidity. Good to know it's not Schism's problem.

@mankeli @esaruoho are you able to build and test latest master to see if MIDI output is steady for you now?

**Comment by esaruoho** [comment]
Cool, I can try midi now.

**Comment by esaruoho** [comment]
well.. midi sounds really shuffly, like every second tick has a SD2 or SD4 on it, delay. so it sounds really shuffly. but _midi notes aren't lost_, so that's great. this kinda loose shuffly timing could be fun for some things.

**Comment by esaruoho** [comment]
also, i'm not able to keyjazz while pattern playback is on, so i can only hear the notes that are played when playback is on. so i can't actually actually play a midi instrument "live", @jangler

**Comment by esaruoho** [comment]
here is a 12/125 example of midi being sent out to outboard synth and sampled. as you can see, it's not "rock solid" ;)
[resample 0004 [2018-01-06 185357].aif.zip](https://github.com/schismtracker/schismtracker/files/1609012/resample.0004.2018-01-06.185357.aif.zip)

<img width="342" alt="screen shot 2018-01-06 at 18 54 48" src="https://user-images.githubusercontent.com/4966687/34641869-166573c6-f313-11e7-947e-e9cc0f5e2d63.png">

**Comment by esaruoho** [comment]
macOS High Sierra, Mid-2015 machine so this should have no issues with midi, per se.

**Comment by esaruoho** [comment]
[resample 0006 [2018-01-06 185736].aif.zip](https://github.com/schismtracker/schismtracker/files/1609016/resample.0006.2018-01-06.185736.aif.zip)
[midi_ticks.it.zip](https://github.com/schismtracker/schismtracker/files/1609017/midi_ticks.it.zip)
here's a .it and it's result. starts with tempo 12, then next 16 steps are tempo 6, then the next 16 steps are tempo 3, then the next 16 steps are tempo 1. as you can see it gets quite choppy.

**Comment by jangler** [comment]
Last time I used MIDI in Schism I didn't have either of those problems (can't check right now), but I'm sending to TiMidity and not to a hardware synth. Don't know if that could make the difference.

**Comment by esaruoho** [comment]
For me, it never made any difference whether I sent MIDI notes from Schism to a secondary DAW via Virtual Midi Cables, or if I sent MIDI notes from Schism to external gear. The drops/shuffliness would always happen.

**Comment by laamaa** [comment]
I experienced really glitchy MIDI timing as well on linux+alsa (external synths via a Edirol UM-2 interface), a lot like in @esaruoho's clip. I first tried hacking all midi related threads to maximum priority, but that didn't affect the situation. At some point I noticed that I was operating in 22050hz for some reason. I tried raising the samplerate in CTRL+F1 menu and it seems setting SR to 96khz seemed to improve the situation a lot, now the MIDI out is at least somewhat usable :)

**Comment by laamaa** [comment]
Now that I played around with MIDI+samples a bit more, it seemed that syncing to audio was broken and midi notes weren't delayed at all? I managed to make a some kind of hack to fix this in my fork, at least I can now use samples+external synths with a RPi 2 https://github.com/laamaa/schismtracker/commit/154426949e360f8fa65789ba22946b2f38ce585b

Edit: Setting the sample rate higher, adding an adjustable MIDI delay compensation setting to MIDI menu (https://github.com/laamaa/schismtracker/commit/de2d020aa7c565aa821bcca77fd4aec1d55da4a6)  and switching to a realtime kernel seemed to solve the problems pretty much for my Pi+internal soundcard+edirol usb midi interface setup
